### ISSUE TITLE
Investigate performance of client-server array transfers

### ISSUE THREAD

**Issue by reuster986** [Bears-R-Us/arkouda#794]
The functions `ak.array()` and `pdarray.to_ndarray()` are about 2x slower on a 128-core/2TB IB system than on Cray XC. Even on XC, the transfer rate for these functions is in the neighborhood of 1 MB/s, which seems quite slow. We should investigate how to improve this performance. The first step is to make a benchmark.

For cases where the client and server share the same filesystem, we could theoretically get a much faster transfer via disk, using `ak.save_all()` and `h5py` on the client side.

**Comment by hokiegeek2** [comment]
@reuster986 @mhmerrill Chapel support for Apache Arrow/Parquet would also be awesome here

**Comment by glitch** [comment]
I started looking into Arrow/Parquet to see how hard it would be to call from Chapel.  Turns out there is a C api and I was able to get a rudimentary example working from Chapel:
https://github.com/glitch/arkouda/wiki/Investigating-Arrow

In theory we could ship this over a socket and read it back out on the python side with `pyarrow`.  It's another set of dependencies, but it doesn't look too bad.

**Comment by hokiegeek2** [comment]
@glitch that's really cool 💯

**Comment by mhmerrill** [comment]
@glitch we have an intern starting on the Chapel team who is going to work on making a Chapel module to wrap [Arrow](http://arrow.apache.org/).
I will put in another issue to track this work.
It would be great if you could look into refactoring the [`src/GenSymIO.chpl`](https://github.com/Bears-R-Us/arkouda/blob/master/src/GenSymIO.chpl), I will also put in another issue to track the refactor.

**Comment by mhmerrill** [comment]
@Bears-R-Us/arkouda-core-dev Since ZMQ doesn't seem to be optimized for data transport... I have asked @hokiegeek2 to look into what library we could use to move the data between the client and the server. Refactoring the client-server messaging into a control/command channel and a data channel. This could also simplify the control/command channel messages to get the binary data out of that path.

**Comment by hokiegeek2** [comment]
👍

**Comment by ronawho** [comment]
As far as speeding up `pdarray.to_ndarray()`itself I think there's probably 2 important aspects here:
 - How fast are we turning the distributed arrays into bytes
 - How quickly are those bytes being transferred across the network

I'm seeing a pretty large dropoff when going from 1->2 locales for turning the arrays into bytes. Looking at the code for this: https://github.com/Bears-R-Us/arkouda/blob/2ce1f6ca670664b9a54973b92f37bdee8b3babfe/src/GenSymIO.chpl#L120-L165

I think we're creating a memory mapped file on locale 0, and writing the distributed array into this. Under the covers I suspect this results in serially iterating over the distributed array and GET'ing individual elements which will be slow, but especially slow on InfiniBand. We should be able to optimize this by creating a local array and in parallel gathering the distributed array using bulk comm. If memory usage is a concern there we can probably do something clever to avoid copying the entire distributed array.

For transferring bytes across the network, this also seems fairly slow to me. This probably depends on ZMQ itself and how large the pipe between the arkouda server and client is. On the InfiniBand system I'm running on I'm not sure if there's an InfiniBand link between the login and the compute nodes, or just a small Ethernet connection. For this it probably makes sense to create a standalone ZMQ benchmark to time just this transfer and see how fast we can get the server->client transfer and then if that can be made fast enough I think it's a question of identifying any other overheads in Arkouda's use of ZMQ.


Here's an ugly standalone benchmark I'm using to explore the client to server bandwidth:


<details>

server.chpl
```chpl
use IO, ZMQ, Time;

config const n = 10**6;

var context: Context;
var socket = context.socket(ZMQ.REP);
socket.bind("tcp://*:5555");

var A:[1..n] int;
var tmpf = openmem();
var tmpw = tmpf.writer(kind=iobig);
tmpw.write(A);

var tmpr = tmpf.reader(kind=iobig, start=0);
var arrayBytes: bytes;
tmpr.readbytes(arrayBytes);

socket.recv(string);

for 1..10 {
  var t: Timer;
  t.start();
  socket.send(arrayBytes);
  socket.recv(string);
  t.stop();
  const sec = t.elapsed();
  const mibPerNode = (n * numBytes(int)):real / (1024.0*1024.0);
  writef("%.2dr MiB/s per node (%.2drs)\n",  mibPerNode/sec, sec);
  stdout.flush();
}
```

client.py
```python
import zmq, time

context = zmq.Context()
socket = context.socket(zmq.REQ)
#socket.connect("tcp://prod-0001:5555")
socket.connect("tcp://amd-0061:5555")

socket.send_string('')

for i in range(10):
    start = time.time()
    message = socket.recv()
    end = time.time()
    socket.send_string('')
    print("Average time = {:.2f} sec".format(end-start))
```

</details>

With this on the 16-node-cs-hdr config I see ~100 MiB/s, which isn't great but isn't horrible. 

http://wiki.zeromq.org/results:10gbe-tests-v432 and http://wiki.zeromq.org/results:ib-tests-v206 have some 10GbE and older InfiniBand results. Based on both of those I would expect higher rates, so it may be worth looking at some pre-existing C/C++ benchmarks to see if this is a chapel/python wrapping issue or just the link speed between login/compute nodes.

**Comment by ronawho** [comment]
> How fast are we turning the distributed arrays into bytes

https://github.com/Bears-R-Us/arkouda/pull/856 improves this by localizing the distributed array before converting to bytes. With that I'm seeing ~25 MB/s for these operations. There's probably a fair amount of overhead here still with writing to and reading from the memory mapped file. We should just be able to use the local arrays buffer directly, but that will involve some hacking and will require communicating the endian-ness instead of forcing big-endian like the current code does.

I'll be curious to see how this new version performs for you in practice and whether you think it's worth continuing to optimize this or if things are fast enough.

**Comment by hokiegeek2** [comment]
@ronawho this looks really cool! 💯

**Comment by ronawho** [comment]
Just so I don't lose it, here's the latest version of the standalone benchmark I've been using:

<details>

server.chpl
```chpl
use IO, ZMQ, Time;
use TestBase, GenSymIO;

writeln(here.name); stdout.flush();
config const n = 10**8;
config const trials = 2;

var context: Context;
var socket = context.socket(ZMQ.REP);
socket.bind("tcp://*:5555");

var st = new owned SymTab();
var A = st.addEntry("A", n, int);

socket.recv(string);
socket.send(trials:string);

for 1..trials {
  socket.recv(string);
  var arrayBytes = tondarrayMsg("tondarrayMsg", "A", st);
  socket.send(arrayBytes);

  var args = socket.recv(bytes).decode();
  var recvArrayBytes = socket.recv(bytes);
  var ret = arrayMsg("arrayMsg", args, recvArrayBytes, st);
  socket.send(ret.msg);
}
```

client.py
```python
import zmq, time, sys, numpy

context = zmq.Context()
socket = context.socket(zmq.REQ)
socket.connect("tcp://{}:5555".format(sys.argv[1]))

socket.send_string('')
trials = int(socket.recv_string())

for i in range(trials):
    start = time.time()
    socket.send_string('')
    message = socket.recv(copy=False).buffer
    A = numpy.frombuffer(message, numpy.int64)
    elapsed = time.time() - start
    mib = len(message) / 1024**2
    print("tondarr time = {:.2f} sec ({:.2f} MiB/s)".format(elapsed, mib/elapsed))

    start = time.time()
    socket.send('int64 {}'.format(A.size).encode(), flags=zmq.SNDMORE)
    socket.send(memoryview(A), copy=False)
    socket.recv_string()
    elapsed = time.time() - start
    print("ak.arr time = {:.2f} sec ({:.2f} MiB/s)\n".format(elapsed, mib/elapsed))
```



</details>

**Comment by ronawho** [comment]
There have been several optimizations to improve the performance of client/server transfers over the last few weeks. On 16-node-xc we're now at around 1700 MiB/s for `to_ndarray` and 900 MiB/s for `ak.array()`. More details below 

---

For `to_ndarray`, initially the server was doing 8-byte GETs from distributed arrays and writing the values into a big-endian memory mapped file, reading that file back into bytes, performing a ZMQ send on that, and on the client side using struct to unpack and then creating a numpy array. This had a lot of inefficiencies so what we do now is localize the distributed array on the server-side using bulk transfer, directly reinterpret the underlying array memory as bytes, send the bytes with ZMQ, and on the client side we now directly use the ZMQ buffer to create a numpy array with `np.frombuffer`.

Additionally, we reduced the number of copies on both the client and the server side by doing things like sending the binary payload separately to avoid copies from concatenation and splitting. There were also some changes made in the Chapel ZMQ shim to reduce copies and parallelize the one copy that remains. And lastly there were some changes contributed to upstream pyzmq to make the ZMQ buffer mutable so we don't have to copy the numpy array to make it mutable.

The table below summarizes the performance impact from various changes on 16-node-xc:

| PR                       | description                             |  to_ndarray | ak.array   |
| ------------------------ | --------------------------------------- | ----------: | ---------: |
| Initial                  | initial perf when this issue was opened |    4 MiB/s  |   5 MiB/s  |
| #856                     | localize server arrays w/ bulk transfer |   21 MiB/s  |  24 MiB/s  |
| #870                     | opt client byte<->array conversions     |   33 MiB/s  |  50 MiB/s  |
| #880                     | opt server byte<->array conversions     |  410 MiB/s  | 175 MiB/s  |
| #885                     | reduce copies on client and server      |  715 MiB/s  | 705 MiB/s  |
| #893                     | eliminate another server copy           |  715 MiB/s  | 760 MiB/s  |
| chapel-lang/chapel#18126 | eliminate copy in Chapel ZMQ shim       |  715 MiB/s  | 830 MiB/s  |
| chapel-lang/chapel#18140 | parallelize remaining Chapel ZMQ copies |  735 MiB/s  | 870 MiB/s  |
| zeromq/pyzmq#1569        | mutable pyzmq buffer, avoid client copy | 1670 MiB/s  | 870 MiB/s  |

Where the chapel changes will be in 1.25.0 (scheduled for release in late September) and the pyzmq changes are in 22.2.0, which was released a few days ago.

I think this is about as fast as we can get things. Technically there is one more copy we could get rid of in our ZMQ shim for `send` but that copy is fast now that we've parallelized it so I don't think it's worth pursuing since it will require some non-trivial API design.

I'm a little confused about why `to_ndarray` is so much faster than `ak.array`. `ak.array` includes creating a distributed array on the server side and that accounts for a little bit of the time, but not all of it. So far what I have found is that on the client side the `zmq.send` seems to be where most of the time is spent. This should be a non-copying send but based on the speed I'm wondering if there is some hidden copy there. Sending things with ZMQ is complicated because the underlying C library wants to take ownership of the buffer being sent. In the Chapel shim we make an extra copy and give ZMQ ownership of this (this is what I was referring to when I said there's one more copy we could get rid of in the previous paragraph) and I'm wondering if this exists on the pyzmq side. I haven't found such a deep copy in the code yet, but it wouldn't surprise me if it's there. Left to my own devices I would probably call it here, but I'm open to digging into that more if you want.

**Comment by ronawho** [comment]
Note that while performance looks great for XC, results on InfiniBand are less stellar at around 100 MiB/s. I opened https://github.com/Bears-R-Us/arkouda/issues/895 to track this.

**Comment by ronawho** [comment]
Offline I believe @reuster986 indicated performance is for array transfers is in good shape, so I think this can be closed now.

**Comment by reuster986** [comment]
Confirmed, and thanks @ronawho !
